File: 003-binary-search/cpp/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Minimal, didactic binary search on an implicit sorted array.
// The array is conceptual: arr[i] = i for 0 <= i < size.

static bool bsearch_ll(long long size, long long target) {
    if (size <= 0) return false;
    long long lo = 0, hi = size - 1;
    while (lo <= hi) {
        long long mid = lo + (hi - lo) / 2;
        if (mid == target) return true;
        if (mid < target) lo = mid + 1; else hi = mid - 1;
    }
    return false;
}

int main(int argc, char** argv) {
    long long size = 100000;
    long long target = size - 1;
    if (argc > 1) size = max(0LL, atoll(argv[1]));
    if (argc > 2) target = max(0LL, atoll(argv[2]));
    cout << boolalpha << "found: " << bsearch_ll(size, target) << "\n";
}

File: 003-binary-search/python/main.py
import sys

# Minimal, didactic binary search over an implicit sorted array.
# The array is conceptual: arr[i] = i for 0 <= i < size.


def bsearch(size: int, target: int) -> bool:
    lo, hi = 0, max(0, size - 1)
    while lo <= hi:
        mid = (lo + hi) // 2
        if mid == target:
            return  True
        if mid < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return False


def main():
    # Args: <size> <target>, defaults: 100000, size-1
    size = 100_000
    target = size - 1
    if len(sys.argv) > 1:
        try:
            size = max(0, int(sys.argv[1]))
            target = size - 1
        except ValueError:
            pass
    if len(sys.argv) > 2:
        try:
            target = max(0, int(sys.argv[2]))
        except ValueError:
            pass
    print(f"found: {bsearch(size, target)}")


if __name__ == "__main__":
    main()

